### ISSUE TITLE
False negative `used-before-assignment` after name defined and used in except

### ISSUE THREAD

**Issue by jacobtylerwalls** [PyCQA/pylint#5780]
### Bug description

This doesn't warn as it should:
```python
try:
  pass
except ValueError:
  msg = 'ha'
  print(msg)
print(msg)  # should raise used-before-assignment
```

But this warns just fine:
```python
try:
  pass
except ValueError:
  msg = 'ha'
print(msg)  # [used-before-assignment]
```

So it's an issue with "consuming" the name too early in the variables checker when used by the first "print(msg)".

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 1.67/10, +8.33)
```

### Expected behavior

************* Module a
a.py:6:6: E0601: Using variable 'msg' before assignment (used-before-assignment)

### Pylint version

```shell
pylint 2.13.0-dev0
astroid 2.9.2
Python 3.10.1 (v3.10.1:2cd268a3a9, Dec  6 2021, 14:28:59) [Clang 13.0.0 (clang-1300.0.29.3)]
```

**Comment by jacobtylerwalls** [comment]
This diff passes the test suite, but two small questions:

- I'd like to make `mark_as_consumed()` itself less brittle so I don't have to do this extra "careful consumption".
- Is it okay to add back this complexity around when to mark as consumed?

```diff
diff --git a/pylint/checkers/variables.py b/pylint/checkers/variables.py
index cbe7b4f1..56942208 100644
--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1371,6 +1371,12 @@ class VariablesChecker(BaseChecker):
         # iterates through parent scopes, from the inner to the outer
         base_scope_type = self._to_consume[start_index].scope_type
 
+        def _careful_consume(nodes_to_delay_consuming):
+            for n in nodes_to_delay_consuming:
+                if n.name in current_consumer.to_consume:
+                    current_consumer.mark_as_consumed(n.name, [n])
+
+        nodes_to_delay_consuming: List[nodes.NodeNG] = []
         for i in range(start_index, -1, -1):
             current_consumer = self._to_consume[i]
 
@@ -1386,13 +1392,20 @@ class VariablesChecker(BaseChecker):
                 # should be added back so that they are marked as used.
                 # They will have already had a chance to emit used-before-assignment.
                 # We check here instead of before every single return in _check_consumer()
-                nodes_to_consume += current_consumer.consumed_uncertain[node.name]
+                # EXCEPTION: nodes in except handlers may not have executed, so delay
+                # marking them as used until exiting the function
+                if utils.get_node_first_ancestor_of_type(node, nodes.ExceptHandler):
+                    nodes_to_delay_consuming += current_consumer.consumed_uncertain[node.name]
+                else:
+                    nodes_to_consume += current_consumer.consumed_uncertain[node.name]
                 current_consumer.mark_as_consumed(node.name, nodes_to_consume)
             if action is VariableVisitConsumerAction.CONTINUE:
                 continue
             if action is VariableVisitConsumerAction.RETURN:
+                _careful_consume(nodes_to_delay_consuming)
                 return
 
+        _careful_consume(nodes_to_delay_consuming)
         # we have not found the name, if it isn't a builtin, that's an
         # undefined name !
         if (
```

**Comment by Pierre-Sassoulas** [comment]
Very similar to #2835 (same false negative but undefined in the other branch)
